src/cata_tiles.cpp

```diff
@@ -1,69 +1,375 @@
+        SDL_CreateRGBSurface(0, tile_width, tile_height, 32, r_mask, g_mask, b_mask, a_mask);
 
-    last_pos_x = 0;
-    last_pos_y = 0;
+    if (!surface) {
+        dbg( D_ERROR ) << "Failed to create surface: " << SDL_GetError();
+    }
+
+    return std::unique_ptr<SDL_Surface> {surface};
+}
+
+bool ends_with(std::string const &s, std::string const &ending)
+{
+    if (ending.size() > s.size()) {
+        return false;
+    }
+
+    return std::equal(ending.rbegin(), ending.rend(), s.rbegin());
+}
+
+//--------------------------------------------------------------------------------------------------
+template <typename Container>
+tile_type const* file_tile_category(
+    Container const &ids, std::string const &id, tile_category const cat, std::string const &sub_cat)
+{
+    long     sym = -1;
+    nc_color col = c_white;
+
+    switch (cat) {
+    case tile_category::none :
+        break;
+    case tile_category::vehicle_part : {
+        // TODO reset rotation here.
+        auto const it = vehicle_part_types.find(id.substr(3));
+        if (it != vehicle_part_types.end()) {
+            if (sub_cat.empty()) {
+                sym = it->second.sym;
+            } else {
+                sym = special_symbol(sub_cat[0]);
+            }
+            col = it->second.color;
+        }
+    } break;
+    case tile_category::terrain : {
+        auto const it = termap.find(id);
+        if (it != termap.end()) {
+            sym = it->second.sym;
+            col = it->second.color;
+        }
+    } break;
+    case tile_category::item :
+        if (auto const it = item_controller->find_template(id)) {
+            sym = it->sym;
+            col = it->color;
+        }
+        break;
+    case tile_category::furniture : {
+        auto const it = furnmap.find(id);       
+        if (it != furnmap.end()) {
+            sym = it->second.sym;
+            col = it->second.color;
+        }
+    } break;
+    case tile_category::trap : {
+        auto const it = trapmap.find(id);
+        if (it != trapmap.end()) {
+            auto const& trap = *traplist[it->second];
+            sym = trap.sym;
+            col = trap.color;
+        }
+    } break;
+    case tile_category::field : {
+        // TODO: field density?    
+        auto const &f = fieldlist[field_from_ident(id)];
+        sym = f.sym;
+        col = f.color[0];
+    } break;
+    case tile_category::lighting :
+        break;
+    case tile_category::monster :
+        // TODO clean up has_mtype
+        if (MonsterGenerator::generator().has_mtype(id)) {
+            const mtype *m = MonsterGenerator::generator().get_mtype(id);
+            int len = m->sym.length();
+            const char *s = m->sym.c_str();
+            sym = UTF8_getch(&s, &len);
+            col = m->color;
+        }
+        break;
+    case tile_category::bullet :
+        break;
+    case tile_category::hit_entity :
+        break;
+    case tile_category::weather :
+        break;
+    default:
```

Comment:
That's a good point -- I think I added it by *default* when I was doing the initial refactoring and didn't go back to reconsider the repercussions. I'll remove the default case to enable the warning to be useful here.